onnx/defs/controlflow/defs.cc

```diff
@@ -261,10 +261,16 @@ void IfInferenceFunction(InferenceContext& ctx) {
             else_elem_type);
       }
 
-      // merge the 'else' shape information to check it's consistent and
-      // augment the 'if' output if possible
-      mergeInShapeInfo(
-          else_output->tensor_type(), *if_output->mutable_tensor_type());
+			if (checkShapeCompatibility(
+              then_output->tensor_type().shape(),
+              else_output->tensor_type().shape())) {
+        // merge the 'else' shape information to check it's consistent and
+        // augment the 'if' output if possible
+        mergeInShapeInfo(
```

Comment:
There are a few issues with this logic. What we need to do is set the resulting shape as the "union" of the shapes produced by the THEN and ELSE branch. I recommend using this function: https://github.com/onnx/onnx/blob/c5af774aa950f6a5cf1a7b5f37d014fb2fbd82bc/onnx/defs/sequence/defs.cc#L21

Reply:
Sure.

Reply:
Using UnionShapeInfo produces wrong shapes in my test. Can I roll back to my original function?

Reply:
Fixed.